datapipelines/src/test/java/io/customer/datapipelines/DeviceAttributesTests.kt

```diff
@@ -0,0 +1,126 @@
+package io.customer.datapipelines
+
+import android.Manifest
+import android.content.pm.PackageManager
+import com.segment.analytics.kotlin.android.plugins.AndroidContextPlugin
+import io.customer.datapipelines.support.core.RobolectricTest
+import io.customer.datapipelines.support.core.TestConfiguration
+import io.customer.datapipelines.support.core.testConfiguration
+import io.customer.datapipelines.support.extensions.deviceToken
+import io.customer.datapipelines.support.extensions.encodeToJsonValue
+import io.customer.datapipelines.support.utils.OutputReaderPlugin
+import io.customer.datapipelines.support.utils.TestConstants
+import io.customer.datapipelines.support.utils.trackEvents
+import io.customer.sdk.extensions.random
+import io.mockk.every
+import kotlinx.serialization.json.booleanOrNull
+import kotlinx.serialization.json.intOrNull
+import kotlinx.serialization.json.jsonPrimitive
+import org.amshove.kluent.shouldBeEmpty
+import org.amshove.kluent.shouldBeEqualTo
+import org.amshove.kluent.shouldContain
+import org.amshove.kluent.shouldHaveSingleItem
+import org.amshove.kluent.shouldHaveSize
+import org.amshove.kluent.shouldNotBeNull
+import org.amshove.kluent.shouldNotBeNullOrBlank
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.annotation.Config
+
+@RunWith(RobolectricTestRunner::class)
+@Config(manifest = Config.NONE)
+class DeviceAttributesTests : RobolectricTest() {
+    private lateinit var outputReaderPlugin: OutputReaderPlugin
+
+    init {
+        // Return permission granted for network state permission so analytics can track network state
+        every { mockApplication.checkCallingOrSelfPermission(Manifest.permission.ACCESS_NETWORK_STATE) } returns PackageManager.PERMISSION_GRANTED
+    }
+
+    @Before
+    override fun setup() {
+        // Keep setup empty to avoid calling super.setup() as it will initialize the SDK
+        // and we want to test the SDK with different configurations in each test
+    }
+
+    override fun setupTestEnvironment(testConfig: TestConfiguration) {
+        super.setupTestEnvironment(testConfig)
+
+        outputReaderPlugin = OutputReaderPlugin()
+        analytics.add(outputReaderPlugin)
+    }
+
+    @Test
+    fun track_givenAutoTrackDeviceAttributesDisabled_expectNoDeviceAttributesInProperties() {
+        setupTestEnvironment(
+            testConfiguration {
+                sdkConfig {
+                    // Disable auto tracking of device attributes
+                    setAutoTrackDeviceAttributes(false)
+                }
+                configurePlugins {
+                    // Add Android context plugin so that device attributes can be tracked by analytics
+                    add(AndroidContextPlugin())
```

Comment:
doesn't it get added by default?

Reply:
Unfortunately not in tests. It is because the plugin is added when using constructor for Android. But since we want to override schedulers with test ones, we have to use Java constructor which by default doesn't add this plugin and we have to include it manually.